codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -69,60 +69,113 @@
+            return handler;
+        }
     }
 
-    public Http2ConnectionHandler(boolean server, boolean validateHeaders) {
-        this(new DefaultHttp2Connection(server), validateHeaders);
-    }
+    public abstract static class BuilderBase<T extends Http2ConnectionHandler, B extends BuilderBase<T, B>> {
+        private Http2Settings initialSettings;
+        private Http2FrameListener listener;
+        private boolean validateHttp2Headers = true;
+        private boolean server = true;
+        private Http2FrameLogger frameLogger;
+        private long gracefulShutdownTimeoutMillis = DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS;
 
-    public Http2ConnectionHandler(Http2Connection connection) {
-        this(connection, true);
-    }
+        @SuppressWarnings("unchecked")
+        public B frameListener(Http2FrameListener listener) {
+            this.listener = listener;
+            return (B) this;
+        }
 
-    public Http2ConnectionHandler(Http2Connection connection, boolean validateHeaders) {
-        this(connection, new DefaultHttp2FrameReader(validateHeaders), new DefaultHttp2FrameWriter());
-    }
+        public final Http2FrameListener frameListener() {
+            return listener;
+        }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2FrameReader frameReader,
-                                  Http2FrameWriter frameWriter) {
-        initialSettings = null;
-        encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader);
-    }
+        @SuppressWarnings("unchecked")
+        public B validateHeaders(boolean validate) {
+            this.validateHttp2Headers = validate;
+            return (B) this;
+        }
 
-    /**
-     * Constructor for pre-configured encoder and decoder. Just sets the {@code this} as the
-     * {@link Http2LifecycleManager} and builds them.
-     */
-    public Http2ConnectionHandler(Http2ConnectionDecoder decoder,
-                                  Http2ConnectionEncoder encoder) {
-        this.initialSettings = null;
-        this.decoder = checkNotNull(decoder, "decoder");
-        this.encoder = checkNotNull(encoder, "encoder");
-        if (encoder.connection() != decoder.connection()) {
-            throw new IllegalArgumentException("Encoder and Decoder do not share the same connection object");
+        @SuppressWarnings("unchecked")
+        public B validateHttp2Headers(boolean validate) {
+            this.validateHttp2Headers = validate;
+            return (B) this;
         }
-    }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2Settings initialSettings) {
-        this(connection, new DefaultHttp2FrameReader(), new DefaultHttp2FrameWriter(),
-                initialSettings);
-    }
+        @SuppressWarnings("unchecked")
+        public B initialSettings(Http2Settings settings) {
+            initialSettings = settings;
```

Comment:
I tend to favor the builder being "dumb" in terms of enforcing constraints, and letting the class that is built determine what is valid and what is not.